### PR TITLE
Allow state of btparse parser to be reset, for parsing multiple files

### PR DESCRIPTION
**Pull request by kwwette**
It is currently very difficult to get Text::BibTeX to parse multiple files, especially when errors are encountered in the first file. This is because btparse has an internal parser state which must be reset between files, but this is currently very hard to accomplish (i.e. bt_parse_entry() must be passed a FILE for which feof() is true, which seems to be difficult to construct from Perl due to buffering, etc.)

This pull request allows the state of `btparse` to be reset explicitly from Perl, so that parsing multiple files becomes possible. As documented in `Changes`, it does the following:
   - bt_parse_entry(): reset parser state if infile == NULL
   - BibTeX.xs: add _reset_parse(), _reset_parse_s() methods to Text::BibTeX::Entry
   - Text::BibTeX::Entry: allow new() or parse() with undefined filehandle; calls _reset_parse()
   - Text::BibTeX::Entry: allow new() or parse_s() with undefined text; calls _reset_parse_s()
   - Text::BibTeX::File: close() calls Text::BibTeX::Entry->new($filename, undef) to reset parser

In short, the parser should now automatically reset for anyone using Text::BibTeX::File to parse BibTeX entries; for anyone using Text::BibTeX::Entry->new($filename, $filehandle) or Text::BibTeX::Entry->new($text) to parse entries, passing $filehandle == undef or $text == undef will also reset the parser.

### CODE REVIEW

**Comment by ambs** [comment]
I need to test this. I used Text::BibTeX to parse multiple files and did not have any issue, so I just need to be sure it still works.

@plk I would appreciate if you test this patch against your biber test suite.

**Comment by plk** [comment]
I did a quick test with these patches and I can confirm that all tests pass fine with biber. I am still not entirely clear what it is designed to solve - would like to have an example file which demonstrates the issue?

**Comment by kwwette** [comment]
Thank you for the testing and comments.

I agree that btparse is currently capable of parsing multiple files, but I think that is provided those files does not contain any errors, and btparse is therefore able to parse each file in its entirety.

My use case is that I'm trying to parse a BibTeX file being edited by the user, which may contain errors, and I want to be able to keep reparsing the file until all errors have been corrected. The problem I found is that when btparse encounters errors during parsing, it is often not be able to parse to the end of the file - say e.g. the user forgot the @ at the start of the BibTeX entry, and so the file now contains non-BibTeX text that cannot be parsed. In this situation, bt_parse_entry() is unable to reset its internal state because end-of-file has not been reached (see btparse/src/input.c, lines 308--322 on the current master). Then once I've given the user a change to correct this error, and try to parse the file again, I get the error on btparse/src/input.c, line 295, because bt_parse_entry() thinks we haven't finished parsing the previous file.

I have tried other ways of solving this problem (e.g. moving the file pointer to end-of-file and calling bt_parse_entry() a final time) but this has not proved to be reliable, perhaps because of Perl file buffering. The most reliable solution would be to have a way to explicitly reset the bt_parse_entry() internal state when a file closes, so that regardless of whether the previous parse was successful or not, btparse is ready to parse a new file.

**Comment by kwwette** [comment]
Is there any further information, tests, etc. required before this pull request is merged? Thanks!

**Comment by ambs** [comment]
Hi. Sorry for the lack of answers. Not lack of interest, just lack of time.
@plk, if you do not have any objection, I'll merge this PR and release a new version.

**Comment by plk** [comment]
Looks fine to me.

**Comment by kwwette** [comment]
@ambs - excellent, thank you very much!

**Comment by ambs** [comment]
We thank you for your work. Sorry if we take too long sometimes :(

**Comment by kwwette** [comment]
No problem at all - and thank you for providing this very useful package!
